src/metabase/search/spec.clj

```diff
@@ -0,0 +1,255 @@
+(ns metabase.search.spec
+  (:require
+   [clojure.set :as set]
+   [clojure.string :as str]
+   [clojure.walk :as walk]
+   [malli.core :as mc]
+   [malli.error :as me]
+   [metabase.config :as config]
+   [metabase.util :as u]
+   [toucan2.core :as t2]))
+
+(def ^:private SearchModel
+  [:enum "dashboard" "table" "dataset" "segment" "collection" "database" "action" "indexed-entity" "metric" "card"])
+
+(def ^:private AttrValue
+  "Key must be present, to show it's been explicitly considered.
+
+  - false: not present
+  - true: given by a column with the same name (snake case)
+  - keyword: given by the corresponding column
+  - vector: calculated by the given expression
+  - map: a sub-select"
+  [:union :boolean :keyword vector? :map])
+
+(def ^:private explicit-attrs
+  "These attributes must be explicitly defined, omitting them could be a source of bugs."
+  [:archived
+   :collection-id
+   :database-id
+   :table-id])
+
+(def ^:private optional-attrs
+  "These attributes may be omitted (for now) in the interest of brevity in the definitions."
+  [:id ; unlike the others, this is implictly true
+   :created-at
+   :creator-id
+   :dataset-query
+   :last-edited-at
+   :last-editor-id
+   :pinned
+   :verified
+   :updated-at])
+
+(def ^:private attr-keys
+  "Keys of a search-model that correspond to concrete columns in the index"
+  (into explicit-attrs optional-attrs))
+
+(def attr-columns
+  "Columns of an ingestion query that correspond to concrete columns in the index"
+  (mapv (comp keyword u/->snake_case_en name) attr-keys))
+
+(assert (not-any? (set explicit-attrs) optional-attrs) "Attribute must only be mentioned in one list")
+
+(def ^:private Attrs
+  (into [:map {:closed true}]
+        (concat (for [k explicit-attrs] [k AttrValue])
+                (for [k optional-attrs] [k {:optional true} AttrValue]))))
+
+(def ^:private NonAttrKey
+  ;; This is rather slow, not great for REPL development.
+  (if config/is-dev?
+    :keyword
+    [:and :keyword [:not (into [:enum] attr-columns)]]))
+
+(def ^:private JoinMap
+  "We use our own schema instead of raw HoneySQL, so that we can invert it to calculate the update hooks."
+  [:map-of :keyword [:tuple :keyword vector?]])
+
+(def ^:private Specification
```

Comment:
can we make this map closed?

Reply:
it's missing :bookmark I believe.

Reply:
Sure - I was just waiting until I figure out stuff like bookmark, which is a bit fuzzy until I figure out some of the new scoring implementation, but that's not a good reason really.